packages/truffle-artifactor/index.js

```diff
@@ -4,97 +4,76 @@ var path = require("path");
 var _ = require("lodash");
 var debug = require("debug")("artifactor");
 
-function Artifactor(destination) {
-  this.destination = destination;
-};
-
-Artifactor.prototype.save = function(object) {
-  var self = this;
-
-  return new Promise(function(accept, reject) {
-    object = Schema.normalize(object);
+class Artifactor {
+  constructor(destination) {
+    this.destination = destination;
+  }
 
-    if (object.contractName == null) {
-      return reject(new Error("You must specify a contract name."));
+  async save(artifactObject) {
+    const self = this;
+
+    const normalizedArtifact = Schema.normalize(artifactObject);
+    const contractName = normalizedArtifact.contractName;
+
+    if (!contractName) throw new Error("You must specify a contract name.");
+
+    const output_path = path.join(self.destination, `${contractName}.json`);
+    let completeArtifact = {};
+
+    // helper for writing artifacts
+    const writeArtifact = _completeArtifact => {
+      _completeArtifact.updatedAt = new Date().toISOString();
+      fs.writeFileSync(
+        output_path,
+        JSON.stringify(_completeArtifact, null, 2),
+        "utf8"
+      );
+    };
+
+    try {
+      const existingArtifact = fs.readFileSync(output_path, "utf8"); // check if artifact already exists
+      const existingArtifactObject = JSON.parse(existingArtifact); // parse existing artifact
+      const normalizedExistingArtifact = Schema.normalize(
+        existingArtifactObject
+      );
+      _.merge(completeArtifact, normalizedExistingArtifact, normalizedArtifact);
+      writeArtifact(completeArtifact);
+    } catch (e) {
+      // if artifact doesn't already exist, write new file
+      if (e.code === "ENOENT") return writeArtifact(normalizedArtifact);
+      else if (e instanceof SyntaxError) throw new Error(e); // catches improperly formatted artifact json
+      throw new Error(e); // catch all other errors
     }
+  }
 
-    var output_path = object.contractName;
-
-    // Create new path off of destination.
-    output_path = path.join(self.destination, output_path);
-    output_path = path.resolve(output_path);
-
-    // Add json extension.
-    output_path = output_path + ".json";
-
-    fs.readFile(output_path, {encoding: "utf8"}, function(err, json) {
-      // No need to handle the error. If the file doesn't exist then we'll start afresh
-      // with a new object.
-
-      var finalObject = object;
-
-      if (!err) {
-        var existingObjDirty;
-        try {
-          existingObjDirty = JSON.parse(json);
-        } catch (e) {
-          reject(e);
-        }
-
-        // normalize existing and merge into final
-        finalObject = Schema.normalize(existingObjDirty);
-
-        // merge networks
-        var finalNetworks = {};
-        _.merge(finalNetworks, finalObject.networks, object.networks);
-
-        // update existing with new
-        _.assign(finalObject, object);
-        finalObject.networks = finalNetworks;
```

Comment:
Careful! I don't think we can delete this separate step to do the network merging.

Reply:
I _think_ I verified this worked fine locally (re-migrating on the same network & also re-migrating onto a diff chain), but sounds like a good candidate for a test.

Reply:
good thinking!